### ISSUE TITLE
Routing and errors

### ISSUE THREAD

**Issue by jkarni** [haskell-servant/servant#224]
Reference for #214 as an issue for cleanliness.

**Comment by jkarni** [comment]
https://github.com/for-GET/http-decision-diagram is a great resource. 

Anyone have thoughts on why 400 should happen before Accept errors? To me this (a) wastes time decoding and validating requests, and (b) makes it unclear whether side-effecting requests with a valid body but unacceptable headers have in fact performed their side-effects.

**Comment by alpmestan** [comment]
If 400 happens *before*, then the body isn't checked is it?

**Comment by jkarni** [comment]
@kosmikus suggested two variants of `:<|>`, one which treats all (non-404 or 405) failures of the LHS as fatal (i.e., RHS doesn't get tried), and one which doesn't. The idea is that in cases where the RHS is `Raw`, we want failures to be fatal so things don't always end up in the `Raw` endpoint, but in other cases both should be tried.

**Comment by jkarni** [comment]
See also #218 for work on this.

**Comment by dredozubov** [comment]
> I think this is a fairly one-sided portrayal; I know @kosmikus had fairly concrete examples for the advantage of having a way of treating non-routing-errors as non-fatal, which would be nice to have here for discussion.

It seems like an issue i mentioned in https://github.com/haskell-servant/servant/pull/214#issuecomment-143232867 is one of them

**Comment by jkarni** [comment]
To keep this discussion updated, kosmikus convinced me that errors should not be fatal on irc. We'll worry about memoReqBody when we have to, I guess

**Comment by jkarni** [comment]
Closed in #218
